### PR TITLE
make io smooth

### LINKED ISSUE #679: curvefs: IO hang a long time when mem cache is full
**Describe the bug (描述bug)**
IO hang a long time when disk cache is full, may be a few seconds.

**To Reproduce (复现方法)**
run vdbench with this profile for several minutes:
fsd=fsd1,anchor=/mnt/whs-gz830,depth=1,width=10,files=10,sizes=(100m,0),shared=yes,openflags=o_direct
fwd=fwd1,fsd=fsd1,threads=10,xfersize=(512,20,4k,20,64k,20,512k,20,1024k,20),fileio=random,fileselect=random,rdpct=50
rd=rd1,fwd=fwd*,fwdrate=max,format=restart,elapsed=7200,interval=1

image

**Expected behavior (期望行为)**
IO not hang.

**Versions (各种版本)**
OS:
Compiler:
curve-mds:
curve-chunkserver:
curve-snapshotcloneserver:
curve-sdk:
nebd:
curve-nbd:

**Additional context/screenshots (更多上下文/截图)**

### PR DESCRIPTION
**Pull request by wuhongsong**
### What problem does this PR solve?

Issue Number: close #679  

Problem Summary:

### What is changed and how it works?

What's Changed:

How it Works:

Side effects(Breaking backward compatibility? Performance regression?): 

### Check List

- [ ] Relevant documentation/comments is changed or added
- [ ] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

**Comment by wuhongsong** [comment]
recheck

**Comment by wuhongsong** [comment]
recheck

File: curvefs/src/client/s3/client_s3_adaptor.cpp [hunk 1]
```diff
@@ -105,10 +107,19 @@ int S3ClientAdaptorImpl::Write(uint64_t inodeId, uint64_t offset,
 
     FileCacheManagerPtr fileCacheManager =
         fsCacheManager_->FindOrCreateFileCacheManager(fsId_, inodeId);
-    if (fsCacheManager_->WriteCacheIsFull()) {
+    ioMtx_.lock();
+    pendingReq_.fetch_add(1, std::memory_order_seq_cst);
+    VLOG(6) << "pendingReq_ is: " << pendingReq_;
+    uint64_t pendingReq = pendingReq_.load(std::memory_order_seq_cst);
+    if ((fsCacheManager_->GetDataCache() + pendingReq*fuseMaxSize_)
```
Comment (ilixiaocui) [inline curvefs/src/client/s3/client_s3_adaptor.cpp:114]: GetDataCacheNumber?

File: curvefs/src/client/s3/client_s3_adaptor.cpp [hunk 2]
```diff
@@ -105,10 +107,19 @@ int S3ClientAdaptorImpl::Write(uint64_t inodeId, uint64_t offset,
...
+      >= fsCacheManager_->GetDataCacheMax()) {
```
Comment (ilixiaocui) [inline curvefs/src/client/s3/client_s3_adaptor.cpp:115]: GetDataCacheMaxNumber?

File: curvefs/conf/client.conf [hunk 3]
```diff
@@ -92,7 +93,7 @@ s3.baseSleepUs=500
 
 # TODO(huyao): use more meaningfull name
 # background thread schedule time
-s3.intervalSec=3
+s3.intervalSec=0
```
Comment (xu-chaojie) [inline curvefs/conf/client.conf:96]: Why change this value to 0

File: curvefs/src/client/s3/client_s3_adaptor.cpp [hunk 4]
```diff
@@ -105,10 +107,19 @@ int S3ClientAdaptorImpl::Write(uint64_t inodeId, uint64_t offset,
...
         LOG(INFO) << "write cache is full, wait flush";
+        fsCacheManager_->DataCacheByteInc(length);
         fsCacheManager_->WaitFlush();
+    } else {
+        fsCacheManager_->DataCacheByteInc(length);
```
Comment (xu-chaojie) [inline curvefs/src/client/s3/client_s3_adaptor.cpp:120]: you can move this line "fsCacheManager_->DataCacheByteInc(length); " out the "if ... else".

File: curvefs/src/client/s3/client_s3_adaptor.cpp [hunk 5]
```diff
@@ -105,10 +107,19 @@ int S3ClientAdaptorImpl::Write(uint64_t inodeId, uint64_t offset,
...
         fsCacheManager_->FindOrCreateFileCacheManager(fsId_, inodeId);
-    if (fsCacheManager_->WriteCacheIsFull()) {
+    ioMtx_.lock();
+    pendingReq_.fetch_add(1, std::memory_order_seq_cst);
+    VLOG(6) << "pendingReq_ is: " << pendingReq_;
+    uint64_t pendingReq = pendingReq_.load(std::memory_order_seq_cst);
+    if ((fsCacheManager_->GetDataCache() + pendingReq*fuseMaxSize_)
+      >= fsCacheManager_->GetDataCacheMax()) {
         LOG(INFO) << "write cache is full, wait flush";
+        fsCacheManager_->DataCacheByteInc(length);
```
Comment (xu-chaojie) [inline curvefs/src/client/s3/client_s3_adaptor.cpp:117]: Why we call DataCacheByteInc here when this request is pendding?

File: curvefs/src/client/s3/client_s3_cache_manager.h [hunk 6]
```diff
@@ -289,6 +289,14 @@ class FsCacheManager {
         return wDataCacheNum_.load(std::memory_order_relaxed);
     }
 
+    uint64_t GetDataCache() {
```
Comment (xu-chaojie) [inline curvefs/src/client/s3/client_s3_cache_manager.h:292]: GetDataCache --> GetDataCacheByte?

File: curvefs/src/client/s3/client_s3_cache_manager.h [hunk 7]
```diff
@@ -289,6 +289,14 @@ class FsCacheManager {
...
+        return wDataCacheByte_.load(std::memory_order_relaxed);
+    }
+
+    uint64_t GetDataCacheMax() {
```
Comment (xu-chaojie) [inline curvefs/src/client/s3/client_s3_cache_manager.h:296]: GetDataCacheMax  -> GetDataCacheMaxSize?

File: curvefs/src/client/s3/client_s3_adaptor.cpp [hunk 8]
```diff
@@ -123,7 +134,10 @@ int S3ClientAdaptorImpl::Write(uint64_t inodeId, uint64_t offset,
         }
     }
     int ret = fileCacheManager->Write(offset, length, buf);
-    VLOG(6) << "write end inodeId:" << inodeId << ",ret:" << ret;
+    pendingReq_.fetch_sub(1, std::memory_order_seq_cst);
+    fsCacheManager_->DataCacheByteDec(length);
```
Comment (xu-chaojie) [inline curvefs/src/client/s3/client_s3_adaptor.cpp:138]: Why we DataCacheByteDec here when fileCacheManager write succeed.

File: curvefs/src/client/s3/client_s3_cache_manager.h (hunk 7 above)
Comment (wuhongsong) [reply curvefs/src/client/s3/client_s3_cache_manager.h:296]: fix

File: curvefs/src/client/s3/client_s3_cache_manager.h (hunk 6 above)
Comment (wuhongsong) [reply curvefs/src/client/s3/client_s3_cache_manager.h:292]: fix

File: curvefs/src/client/s3/client_s3_adaptor.cpp (hunk 2 above)
Comment (wuhongsong) [reply curvefs/src/client/s3/client_s3_adaptor.cpp:115]: fix

File: curvefs/src/client/s3/client_s3_adaptor.cpp (hunk 1 above)
Comment (wuhongsong) [reply curvefs/src/client/s3/client_s3_adaptor.cpp:114]: fix

File: curvefs/src/client/s3/client_s3_adaptor.cpp [hunk 9]
```diff
@@ -105,10 +107,19 @@ int S3ClientAdaptorImpl::Write(uint64_t inodeId, uint64_t offset,
 
     FileCacheManagerPtr fileCacheManager =
         fsCacheManager_->FindOrCreateFileCacheManager(fsId_, inodeId);
-    if (fsCacheManager_->WriteCacheIsFull()) {
+    ioMtx_.lock();
```
[the remark this answers is not in the document]
Comment (wuhongsong) [reply curvefs/src/client/s3/client_s3_adaptor.cpp:110]: fix

File: curvefs/conf/client.conf (hunk 3 above)
Comment (wuhongsong) [reply curvefs/conf/client.conf:96]: this feature maybe disable later by huyao

File: curvefs/src/client/s3/client_s3_adaptor.cpp (hunk 8 above)
Comment (wuhongsong) [reply curvefs/src/client/s3/client_s3_adaptor.cpp:137]: just release DataCacheByteInc before.

File: curvefs/src/client/s3/client_s3_adaptor.cpp (hunk 5 above)
Comment (wuhongsong) [reply curvefs/src/client/s3/client_s3_adaptor.cpp]: all pending req will calculate prevent from mem cache exceed s3.writeCacheMaxByte

File: curvefs/src/client/s3/client_s3_adaptor.cpp (hunk 4 above)
Comment (wuhongsong) [reply curvefs/src/client/s3/client_s3_adaptor.cpp:115]: fix

File: curvefs/conf/client.conf [hunk 10]
```diff
@@ -81,6 +81,7 @@ volume.fsBlockSize=4096
 #### s3
 s3.blocksize=1048576
 s3.chunksize=4194304
+s3.fuseMaxSize=131072
```
[the remark this answers is not in the document]
Comment (wuhongsong) [reply curvefs/conf/client.conf:85]: fix

File: curvefs/test/client/client_s3_adaptor_test.cpp [hunk 11]
```diff
@@ -176,7 +176,7 @@ TEST_F(ClientS3AdaptorTest, test_first_write_2) {
     S3ClientAdaptorImpl* s3ClientAdaptor2;
     S3ClientAdaptorOption option2;
     option2.nearfullRatio = 100;
-    option2.writeCacheMaxByte = 104857600;
+    option2.writeCacheMaxByte = 10485760000;
```
[the remark this answers is not in the document]
Comment (wuhongsong) [reply curvefs/test/client/client_s3_adaptor_test.cpp]: fix

**Review by xu-chaojie — APPROVED** [review]

**Review by ilixiaocui — APPROVED** [review]

**Pull request merged**